### ISSUE TITLE
Deprecated features get removed from c++20 in MSVC

### ISSUE THREAD

**Issue by maxGimeno** [CGAL/cgal#5020]
## Issue Details
(Taken from http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2131r0.html#removed)
Here is a list of all that is being removed from the Preview version of MSVC2019. We will have to fix all of this to properly support c++20.

### Removed in C++20
 - [x] the `noexcept`-specifier `throw()` (use `noexcept` instead)
 - [x] functional adaptors `not1`, `not2`, `unary_negate`, `binary_negate`, as well as the class members `result_type`, `argument_type`, `first_argument_type`, `second_argument_type` (the functionality of `bind` and `not_fn` is strictly superior).
 - [x] redundant members of allocator (which are subsumed by the allocator_traits defaults): `pointer`, `const_pointer`, `reference`, `const_reference`, `rebind`, `address`, `construct`, `destroy`.
 - [x] traits: `is_literal_type` and `result_of`.

### Deprecated since C++20
 - [x] The capture of `this` via `[=]` is now deprecated, since it is irregular and potentially confusing with `[=, *this]`. The capture should now be spelled `[=, this]`.

**Comment by Vishal19111999** [comment]
Hey, I am a beginner ! Can I work on the first one? I mean using noexcept() specifier instead of throw()?

**Comment by Vishal19111999** [comment]
@maxGimeno There are no throw() specifiers in the code

**Comment by maxGimeno** [comment]
good then

**Comment by Vishal19111999** [comment]
@maxGimeno I am a beginner, can I work on fixing capture of "this"?I see 3 places with the usage of [=]

**Comment by maxGimeno** [comment]
Yes of course, do proceed

**Comment by Vishal19111999** [comment]
@maxGimeno the build (which was using C++ 14) failed in my PR #5059 as [=,this] was introduced in [C++ 17](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0409r2.html), On my machine the build was successful using C++ 20 which deprecates [=]. Please tell what to do about it?

**Comment by Vishal19111999** [comment]
@maxGimeno I have reverted it back for now and working on functional adapters instead! I think this is not the current time to remove [=], guide me if I am wrong

**Comment by maxGimeno** [comment]
You can use #ifdef CGAL_CXX17 to distinguish each case.

**Comment by Vishal19111999** [comment]
@maxGimeno I needed CGAL_CXX20 which was  not defined, so I defined it in this PR

**Comment by maxGimeno** [comment]
Why do tou need CXX_20 ?

**Comment by Vishal19111999** [comment]
@maxGimeno for setting [=] to [=,this]

[=,this] is valid only from c++ 20 but [=] is deprecated in C++ 20

I have put links in above comments

**Comment by maxGimeno** [comment]
Ah ok, sorry, I thought it was introduced in c++17 but it was the year 2017. My bad

**Comment by lrineau** [comment]
> @maxGimeno for setting [=] to [=,this]
> 
> [=,this] is valid only from c++ 20 but [=] is deprecated in C++ 20
> 
> I have put links in above comments

You are wrong. The capture `[=, this]` is valid since C++11, and has always the same meaning: it captures `this` by value (in the sense that the pointed object is captured by reference). What was introduced in C++17 is the capture of `*this`, that captures the object pointed by `this` by value.

If you capture with `[&]` or `[=]`, then the object `*this` is implicitly captured by reference (since C++11). But that was surprising because `[=]` means "capture by copy". So the *implicit* capture of `*this` by reference was deprecated in C++20: the user has to chose to capture *explicitly* `this` or `*this` (by the later was introduced in C++17.

See https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture.

**Comment by lrineau** [comment]
> > @maxGimeno for setting [=] to [=,this]
> > [=,this] is valid only from c++ 20 but [=] is deprecated in C++ 20
> > I have put links in above comments
> 
> You are wrong. The capture `[=, this]` is valid since C++11, and has always the same meaning: it captures `this` by value (in the sense that the pointed object `*this` is captured by reference). What was introduced in C++17 is the capture of `*this`, that captures the object pointed by `this` by value.
> 
> If you capture with `[&]` or `[=]`, then the object `*this` is implicitly captured by reference (since C++11). But that was surprising because `[=]` means "capture by copy". So the _implicit_ capture of `*this` by reference was deprecated in C++20: the user has to chose to capture _explicitly_ `this` or `*this` (by the later was introduced in C++17.
> 
> See [en.cppreference.com/w/cpp/language/lambda#Lambda_capture](https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture).

As @maxGimeno pointed out IRL, I am wrong: the page I quote says:
> If the capture-default is `=`, subsequent simple captures must begin with `&` or be `*this` (since C++17) or `this` (since C++20). So you were right `[=, this]` is only valid since C++20. If I try otherwise, the clang compiler says so:
```
test.cpp:3:16: warning: explicit capture of 'this' with a capture default of '=' is a C++20 extension [-Wc++20-extensions]
    return [=, this] {};
               ^
```

**Comment by Vishal19111999** [comment]
@lrineau lrineau I've updated my PR #5060 -> fixes [=] now

**Comment by lrineau** [comment]
I reopen, because #5060 was only a partial fix.
